# Part 1 - Products Consistency Checks

In [ ]:
# Importing libraries

import pandas as pd
import numpy as np
import os

In [4]:
# Creating products data frame

path = r"C:\Users\David\Desktop\CareerFoundry\Achievement 4\Instacart Basket Analysis"

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

In [5]:
# Checking for null values in df_prods

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [6]:
# Creating a subset data frame of df_prods where the "product_name" column contains null values

df_nan = df_prods[df_prods.product_name.isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [7]:
# Verifying the shape of df_prods

df_prods.shape

(49693, 5)

In [8]:
# Creating a df_prods_clean Data Frame that eliminates null values from the 'product_name' column

df_prods_clean = df_prods[df_prods.product_name.isnull() == False]
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [9]:
# Verifying shape of df_prods_clean Data Frame

df_prods_clean.shape

(49677, 5)

In [10]:
# Creating a Data Frame from a subset of df_prods_clean that contains duplicates

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [11]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [12]:
# Creating a data frame that removes duplicates from df_prods_clean

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [13]:
# Verifying the shape of new Data Frame

df_prods_clean_no_dups.shape

(49672, 5)

In [100]:
# Exporting clean data frame to 'products_checked.csv'

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))



# Part 2 - Data Consistency Checks Tasks

In [15]:
#Run the df.describe() function on your df_prods dataframe

df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [16]:
# Using your new knowledge about how to interpret the output of this function, share in a markdown cell 
#whether anything about the data looks off or should be investigated further

The max in the "prices" column is much higher than expected at 99999.

In [87]:
# Check for mixed-type data in your df_ords dataframe

# Creating df_ords Data Frame
path = r"C:\Users\David\Desktop\CareerFoundry\Achievement 4\Instacart Basket Analysis"
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

# Creating check for mixed_type data by columns

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [89]:
# Run a check for missing values in your df_ords dataframe.
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [90]:
# In a markdown cell, report your findings and propose an explanation for any missing values you find.

The only column with missing values was in the days_since_prior_order column. More than likely these values are null due to there being no record of a prior order, indicating that the order is for a first-time user. Because there are too many records to ignore or delete, we could create a column indicating whether the order is for a first-time user.

In [91]:
# Removing unnamed column from df_ords

columns = df_ords.columns.to_list()
df_ords = df_ords[columns[1:]]
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [93]:
# Address the duplicates using an appropriate method

df_ords['new_user'] = df_ords.days_since_prior_order.isnull()
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  new_user  
0                     NaN      True  
1                    15.0     False  
2                    21.0     False  
3                    29.0     False  
4                    28.0     False

In [85]:
# In a markdown cell, explain why you used your method of choice.

By creating a flag that identifies a new user it enables us to access the entirity of the data frame without deleting any records. In this case, missing values in the 'days_since_prior_order' column is still valuable information worth analyzing.

In [95]:
# Run a check for duplicate values in your df_ords data.

df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, new_user]
Index: []

In [98]:
# In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

There were no duplicate values in the df_ords Data Frame.

In [99]:
# Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder 
# and give them appropriate, succinct names.

df_prods.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_clean'))
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_clean')) 